pkg/proxy/ipvs/proxier.go

```diff
@@ -2069,6 +2069,12 @@ func (proxier *Proxier) syncEndpoint(svcPortName proxy.ServicePortName, onlyNode
 				newEndpoints = localReadyTerminatingEndpoints
 			}
 		}
+
+		// Insert non-local endpoints to mimic the behavior of the iptables proxier
+		// https://github.com/kubernetes/kubernetes/issues/93456
+		if len(newEndpoints) == 0 {
+			newEndpoints = readyEndpoints
```

Comment:
(For readers - this came up at sig-net call today :)

> This is like implementing "PreferLocal" semantics specific to IPVS 

My thoughts exactly, and I think you know my feeling on PreferLocal.

Why do people set `ETP=local`?
1) To ensure the preservation of client IP
2) To optimize (avoid the "2nd hop")

This PR breaks both guarantees in a non-deterministic way, for all clients.  I don't think that's OK.

_IF_ we could do this ONLY for traffic that orginates on "this" node, would it be OK?  You'd still be guaranteeing (1) and there is no "2nd hop", so (2) is satisfied.  I can see how to express this in iptables, but I don't know how to do it in IPVS - is it possible?

Note, there are 2 ways to spin this:

Assume a service with a NodePort and ETP=Local.

1) Traffic from a client pod to the nodeport on that pod's node may/should/must(?) use non-local endpoints when local endpoints are not available.  Such traffic is still considered "external", but this special case is afforded.

2) Traffic from a client pod to the nodeport on that pod's node is considered "internal" and is subject to ITP rather than ETP.  Traffic from a client pod to the nodeport on a different node is considered "external" and is subject to ETP.

In either case, we are ONLY discussing traffic from a client pod to a nodeport _on that pod's node_.  Other traffic (pod->other-node or non-pod->any-node) is "external" and must respect ETP.

Reply:
@danwinship ICYMI

Reply:
> This PR breaks both guarantees in a non-deterministic way, for all clients.

How?

On nodes _with_ local endpoints there is _no difference at all_, so nothing is broken.

On nodes _without_ local endpoints "other" endpoints are inserted, which means:

1. From a POD on a node without endpoints: This will break 2) (obviously) but not 1). The (pod) source is preserved.
2. From main netns on a node (including hostNetwork:true): This does not work, please see https://github.com/kubernetes/kubernetes/issues/93456#issuecomment-733494978
3. External traffic: traffic arriving on a node with endpoints works as usual (no difference). Traffic arriving on a node _without_ endpoints will not work but "in a roundabout and inefficient way" https://github.com/kubernetes/kubernetes/issues/93456#issuecomment-733069629

> The most common use-case is when a service is accessed from PODs but the service may be local or on another cluster. The application want's to access the service in an uniform way using it's external (loadBalancer) IP. But for services with externalTrafficPolicy:Local this will fail if no server is executing on the local node.

The use-case is not uncommon and is a reason in several cases to _not_ use ipvs. Proxy-mode=iptables has this function without an API-driven way to allow it and IMHO ipvs should work in a similar way.

#### Real use-case

The reason I was asked to bring this up was an application that had the problem described above but _also_ wanted to use source based policy routing to direct return traffic from a loadBalancerIP to a specific interface on the node (in telco there may be very many interfaces). But that _only_ works with proxy-mode=ipvs! So they were stuck.

Reply:
Maybe I misunderstand the patch - this seems to apply to both traffic from pods but also actual external traffic.  Is that wrong?

> External traffic: traffic arriving on a node with endpoints works as usual (no difference). Traffic arriving on a node without endpoints will not work but "in a roundabout and inefficient way"

I see.  If we don't SNAT it, then yes we are setting it up to fail, which is what we want.  So as you say "roundabout and inefficient".  Hmm.  Super subtle that is.

> I can see how to express this in iptables, but I don't know how to do it in IPVS - is it possible?

I guess the IPVS expression is "don't SNAT".  But don't we always use IPVS masquerade mode?

Reply:
> But don't we always use IPVS masquerade mode?

Yes, but the meaning in ipvs is different from iptables, it means DNAT. The VIP address is NAT'ed to the pod address. The SNAT (masquerade) is made by iptables rules even for proxy-mode=ipvs.

For ETP=local the SNAT is bypassed with a match to the `KUBE-LOAD-BALANCER-LOCAL` ipset;
```
Chain KUBE-LOAD-BALANCER (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    1    60 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            /* Kubernetes service load balancer ip + port with externalTrafficPolicy=local */ match-set KUBE-LOAD-BALANCER-LOCAL dst,dst
    0     0 KUBE-MARK-MASQ  all  --  *      *       0.0.0.0/0            0.0.0.0/0           
```